### PR TITLE
add Level Zero memory provider

### PR DESCRIPTION
**Pull request by bratpiorka**
This PR adds a Level Zero memory provider that supports #225 enhancement. TODO:
- [x] Make Level Zero provider a separate lib that dynamically loads Level Zero functions from ze_loader
- [x] Create a separate workflow for this with required deps (Level Zero) that will run on GPU
- [x] Add basic test with simple allocation/deallocation
- [x] Update README and docs

Please note that shared memory would be supported in separate PR

### CODE REVIEW

**Comment by igchor** [comment]
@bratpiorka can you describe what's your plan regarding integration with UR? Do we want to have a duplicate provider in UMF for now or are you aiming to use this provider in UR? If it's the latter, there are more features you need to support: see the original implementation here https://github.com/oneapi-src/unified-runtime/blob/main/source/adapters/level_zero/usm.cpp#L170

**Comment by igchor** [comment]
Also, the L0 provider should be a separate static library: see https://github.com/oneapi-src/unified-memory-framework/issues/225

**Comment by bratpiorka** [comment]
@igchor thanks for the links. What do you think about creating a basic functionality first and then adding more features (in separate PRs) so it could finally replace the providers used in UR?

**Comment by igchor** [comment]
> @igchor thanks for the links. What do you think about creating a basic functionality first and then adding more features (in separate PRs) so it could finally replace the providers used in UR?

Sure, that sounds reasonable, I just wanted to highlight that there might be some gaps with the current implementation.

File: include/umf/providers/provider_ze_memory.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+#endif
+
+/// @brief ze memory provider settings struct
+typedef struct ze_memory_provider_params_t {
+    ze_context_handle_t context;
+    ze_device_handle_t device;
+    ze_device_mem_alloc_desc_t device_mem_alloc_desc;
+} ze_memory_provider_params_t;
+
+umf_memory_provider_ops_t *umfZeMemoryProviderOps(void);
```
Comment (igchor) [inline include/umf/providers/provider_ze_memory.h:26]: I think it should be called umfLevelZeroMemoryProvider (file as well). It would be more consistent with UR naming.

File: src/libumf.map [hunk 2]
```diff
@@ -35,6 +35,7 @@ UMF_1.0 {
         umfMemoryProviderCreateFromMemspace;
         umfMemspaceCreateFromNumaArray;
         umfMemspaceDestroy;
+        umfZeMemoryProviderOps;
```
Comment (igchor) [inline src/libumf.map:38]: What about windows .map?

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -58,6 +58,14 @@ if(UMF_BUILD_OS_MEMORY_PROVIDER)
     endif()
 endif()
 
+if(UMF_BUILD_ZE_MEMORY_PROVIDER)
+    set(UMF_SOURCES_LINUX ${UMF_SOURCES_LINUX}
+        provider/provider_ze_memory.c)
+    if(LINUX)
+        set(UMF_LIBS ${UMF_LIBS} ze_loader)
```
Comment (igchor) [inline src/CMakeLists.txt:65]: We should either:
- look for L0 using find_package() or similar
- use FetchContent to download L0 header (UR does that)

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -58,6 +58,14 @@ if(UMF_BUILD_OS_MEMORY_PROVIDER)
     endif()
 endif()
 
+if(UMF_BUILD_ZE_MEMORY_PROVIDER)
+    set(UMF_SOURCES_LINUX ${UMF_SOURCES_LINUX}
+        provider/provider_ze_memory.c)
+    if(LINUX)
```
Comment (igchor) [inline src/CMakeLists.txt:64]: Why is this linux-specific? L0 also works on windows

File: src/provider/provider_ze_memory.c [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        (ze_memory_provider_t *)malloc(sizeof(ze_memory_provider_t));
+    if (!ze_provider) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    ze_provider->context = ze_params->context;
+    ze_provider->device = ze_params->device;
+    ze_provider->device_mem_alloc_desc = ze_params->device_mem_alloc_desc;
+
+    assert(ze_provider->device_mem_alloc_desc.stype ==
```
Comment (igchor) [inline src/provider/provider_ze_memory.c:40]: Return INVALID_ARGUMENT here?

File: CMakeLists.txt [hunk 6]
```diff
@@ -8,6 +8,7 @@ project(unified-memory-framework VERSION 0.1.0 LANGUAGES C)
 # Build Options
 option(UMF_BUILD_SHARED_LIBRARY "Build UMF as shared library" OFF)
 option(UMF_BUILD_OS_MEMORY_PROVIDER "Build OS memory provider" ON)
+option(UMF_BUILD_ZE_MEMORY_PROVIDER "Build Level Zero memory provider" ON)
```
Comment (igchor) [inline CMakeLists.txt:11]: Default value (ON) does not match the README

**Comment by vinser52** [comment]
> @igchor thanks for the links. What do you think about creating a basic functionality first and then adding more features (in separate PRs) so it could finally replace the providers used in UR?

But what is the issue of migrating the whole UR implementation as is? Probably with some minor tweaks

**Comment by vinser52** [comment]
> Also, the L0 provider should be a separate static library: see #225

@igchor did you mean that L0 provider should be only static library or both: static and dynamic?

**Comment by bratpiorka** [comment]
> > @igchor thanks for the links. What do you think about creating a basic functionality first and then adding more features (in separate PRs) so it could finally replace the providers used in UR?
> 
> But what is the issue of migrating the whole UR implementation as is? Probably with some minor tweaks

This version of the Level Zero provider is very simple and is intended to be used by oneCCL, but does not contain any oneCCL-specific code. On the other hand, the UR provider contains a lot of UR-specific code and is quite complex. Making it generic would require a lot of work. Also keep in mind that most of the contents of this PR are additional files needed by umf: workflows, cmake files, def/map etc. So even with this PR porting is still an option - it wouldn't change a lot of code.

**Comment by igchor** [comment]
> > Also, the L0 provider should be a separate static library: see #225
> 
> @igchor did you mean that L0 provider should be only static library or both: static and dynamic?

I think we need the possibility to build it as a static library for UR. As for dynamic library, I don't know at this moment.

**Comment by vinser52** [comment]
> > > @igchor thanks for the links. What do you think about creating a basic functionality first and then adding more features (in separate PRs) so it could finally replace the providers used in UR?
> > 
> > 
> > But what is the issue of migrating the whole UR implementation as is? Probably with some minor tweaks
> 
> This version of the Level Zero provider is very simple and is intended to be used by oneCCL, but does not contain any oneCCL-specific code. On the other hand, the UR provider contains a lot of UR-specific code and is quite complex. Making it generic would require a lot of work. Also keep in mind that most of the contents of this PR are additional files needed by umf: workflows, cmake files, def/map etc. So even with this PR porting is still an option - it wouldn't change a lot of code.

My point is since UR already uses UMF and we agreed to move L0 provider impl from UR repo to UMF we should migrate it first (of course UR specific code should be cleaned up) and make sure we are not breaking UR flow. And then we can test if it fits for oneCCL needs.

**Comment by vinser52** [comment]
> > > Also, the L0 provider should be a separate static library: see #225
> > 
> > 
> > @igchor did you mean that L0 provider should be only static library or both: static and dynamic?
> 
> I think we need the possibility to build it as a static library for UR. As for dynamic library, I don't know at this moment.

Will it introduce any runtime dependencies to some L0 components? It is important in context of MPI, today they have no link dependency to L0 they try to load L0 at runtime (using dlopen) if I am correct. I will double-check.

**Comment by bratpiorka** [comment]
> > > > @igchor thanks for the links. What do you think about creating a basic functionality first and then adding more features (in separate PRs) so it could finally replace the providers used in UR?
> > > 
> > > 
> > > But what is the issue of migrating the whole UR implementation as is? Probably with some minor tweaks
> > 
> > 
> > This version of the Level Zero provider is very simple and is intended to be used by oneCCL, but does not contain any oneCCL-specific code. On the other hand, the UR provider contains a lot of UR-specific code and is quite complex. Making it generic would require a lot of work. Also keep in mind that most of the contents of this PR are additional files needed by umf: workflows, cmake files, def/map etc. So even with this PR porting is still an option - it wouldn't change a lot of code.
> 
> My point is since UR already uses UMF and we agreed to move L0 provider impl from UR repo to UMF we should migrate it first (of course UR specific code should be cleaned up) and make sure we are not breaking UR flow. And then we can test if it fits for oneCCL needs.



> > > > Also, the L0 provider should be a separate static library: see #225
> > > 
> > > 
> > > @igchor did you mean that L0 provider should be only static library or both: static and dynamic?
> > 
> > 
> > I think we need the possibility to build it as a static library for UR. As for dynamic library, I don't know at this moment.
> 
> Will it introduce any runtime dependencies to some L0 components? It is important in context of MPI, today they have no link dependency to L0 they try to load L0 at runtime (using dlopen) if I am correct. I will double-check.

oneCCL also loads L0 at runtime

**Comment by igchor** [comment]
> > > > Also, the L0 provider should be a separate static library: see #225
> > > 
> > > 
> > > @igchor did you mean that L0 provider should be only static library or both: static and dynamic?
> > 
> > 
> > I think we need the possibility to build it as a static library for UR. As for dynamic library, I don't know at this moment.
> 
> Will it introduce any runtime dependencies to some L0 components? It is important in context of MPI, today they have no link dependency to L0 they try to load L0 at runtime (using dlopen) if I am correct. I will double-check.

@vinser52  Okay, then in that case we'll need a dynamic version I guess. Can you add this info to the issue so that we don't miss it?

**Comment by bratpiorka** [comment]
> > > > @igchor thanks for the links. What do you think about creating a basic functionality first and then adding more features (in separate PRs) so it could finally replace the providers used in UR?
> > > 
> > > 
> > > But what is the issue of migrating the whole UR implementation as is? Probably with some minor tweaks
> > 
> > 
> > This version of the Level Zero provider is very simple and is intended to be used by oneCCL, but does not contain any oneCCL-specific code. On the other hand, the UR provider contains a lot of UR-specific code and is quite complex. Making it generic would require a lot of work. Also keep in mind that most of the contents of this PR are additional files needed by umf: workflows, cmake files, def/map etc. So even with this PR porting is still an option - it wouldn't change a lot of code.
> 
> My point is since UR already uses UMF and we agreed to move L0 provider impl from UR repo to UMF we should migrate it first (of course UR specific code should be cleaned up) and make sure we are not breaking UR flow. And then we can test if it fits for oneCCL needs.

I looked into the code more deeply + consulted with @igchor and we could break the existing L0 provider in UR into two providers: a UR-specific top-level provider (implemented in UR) and a low-level generic L0 provider that just translates UMF provider API to ZE calls. With this approach, the provider implementation in this PR should be really simple and generic so it could be used by both UR and oneCCL

**Comment by vinser52** [comment]
> > > > > Also, the L0 provider should be a separate static library: see #225
> > > > 
> > > > 
> > > > @igchor did you mean that L0 provider should be only static library or both: static and dynamic?
> > > 
> > > 
> > > I think we need the possibility to build it as a static library for UR. As for dynamic library, I don't know at this moment.
> > 
> > 
> > Will it introduce any runtime dependencies to some L0 components? It is important in context of MPI, today they have no link dependency to L0 they try to load L0 at runtime (using dlopen) if I am correct. I will double-check.
> 
> @vinser52 Okay, then in that case we'll need a dynamic version I guess. Can you add this info to the issue so that we don't miss it?

Sure, I will update the issue. But, the dynamic version can be omitted if our static version will try to dlopen/dlsym `libze_loader.so` - In that case we are not making any hard dependency.

File: include/umf/providers/provider_ze_memory.h (hunk 1 above)
Comment (bratpiorka) [reply include/umf/providers/provider_ze_memory.h]: changed

File: src/CMakeLists.txt (hunk 3 above)
Comment (bratpiorka) [reply src/CMakeLists.txt]: I changed this and ze_loader is loaded in runtime using dlopen

File: src/provider/provider_ze_memory.c (hunk 5 above)
Comment (bratpiorka) [reply src/provider/provider_ze_memory.c]: done

File: CMakeLists.txt (hunk 6 above)
Comment (bratpiorka) [reply CMakeLists.txt]: done

File: .github/workflows/gpu.yml [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        with:
+          repository: oneapi-src/level-zero
+          path: level_zero
+  
+      - name: Build and Install Level Zero
+        run: |
+            cd level_zero
+            mkdir build
+            cd build
+            cmake .. -DCMAKE_INSTALL_PREFIX=/usr
```
Comment (lukaszstolarczuk) [inline .github/workflows/gpu.yml:42]: don't install in `/usr` if on self-hosted runner 😉 pls install locally, somewhere in workspace

File: .github/workflows/gpu.yml [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+            sudo cmake --build . --config Release --target install
```
Comment (lukaszstolarczuk) [inline .github/workflows/gpu.yml:43]: sudo won't be needed

**Comment by pbalcer** [comment]
> > 
> 
> Perhaps, but I think UR might prefer to have a static library that doesn't perform any dlopens - at least for the case where all dependencies are linked statically. @pbalcer is that true? or is there no difference for UR?

UR would require two variants of this provider: one that is fully dynamically linked that uses the exact same L0 symbols as the adapter (so not dlopen'ed), and one that uses static linking and then uses symbols that from the library that the adapter already dlopen'ed (so we'd probably need to dlopen the exact same library path).

**Comment by bratpiorka** [comment]
> Let's discus it on our sync on Monday. I doubt if we need dlopen at all. maybe just dlsym is enough. My thoughts are the following:
> 
> * L0 memory provider is used only when client already uses L0.
> * So L0 should already be loaded into the process.
> * UMF should just find the corresponding  L0 symbols that are already loaded in the process somehow. It should be possible with `dlsym` and `RTLD_DEFAULT` (a special pseudo-handle)
> 
> With such schema, clients will use L0 in the same way like today (explicit linking, dlopen, etc.). UMF will not have dependency to `libze_loader`. L0 memory provider implementation will try to find corresponding symbols at runtime.

@vinser52 I made changes that follow this algorithm: first I try to get symbols using NULL handle and only if this fails I try to load libze_loader.so

**Comment by bratpiorka** [comment]
@pbalcer I will resolve all UR-related issues in a separate PR when I also try to do a PR in UR repo that integrates with this provider

**Comment by igchor** [comment]
> > > > 
> > > 
> > > 
> > > Perhaps, but I think UR might prefer to have a static library that doesn't perform any dlopens - at least for the case where all dependencies are linked statically. @pbalcer is that true? or is there no difference for UR?
> > 
> > 
> > UR would require two variants of this provider: one that is fully dynamically linked that uses the exact same L0 symbols as the adapter (so not dlopen'ed), and one that uses static linking and then uses symbols that from the library that the adapter already dlopen'ed (so we'd probably need to dlopen the exact same library path).
> 
> Let's discus it on our sync on Monday. I doubt if we need dlopen at all. maybe just dlsym is enough. My thoughts are the following:
> 
> * L0 memory provider is used only when client already uses L0.
> * So L0 should already be loaded into the process.
> * UMF should just find the corresponding  L0 symbols that are already loaded in the process somehow. It should be possible with `dlsym` and `RTLD_DEFAULT` (a special pseudo-handle)
> 
> With such schema, clients will use L0 in the same way like today (explicit linking, dlopen, etc.). UMF will not have dependency to `libze_loader`. L0 memory provider implementation will try to find corresponding symbols at runtime.

@vinser52 @pbalcer  But keep in mind that what Rafal implemented is using dlopen/dlsym on level zero **loader**. The loader itself will use dlopen to load appropriate drivers (from what I understand). So I don't think we need to worry about what symbols are being used, do we? Btw, assuming there would be a static version of L0 loader (could we ask L0 team?), could we just link with that and rely on L0 discovery instead of doing any dlopen/dlsym ourselves?

File: test/CMakeLists.txt [hunk 9]
```diff
@@ -124,6 +125,12 @@ if(UMF_BUILD_OS_MEMORY_PROVIDER AND LINUX) # OS-specific functions are implement
                  LIBS umf_utils numa)
 endif()
 
+if(UMF_BUILD_LIBUMF_PROVIDER_LEVEL_ZERO AND LINUX) # TODO
```
Comment (igchor) [inline test/CMakeLists.txt:128]: Why TODO? Is there anything missing to support it on windows?

File: test/providers/provider_level_zero.cpp [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+    umf_memory_provider_handle_t provider = nullptr;
+    umf_result_t umf_result = umfMemoryProviderCreate(
+        umfLevelZeroMemoryProviderOps(), &params, &provider);
+    ASSERT_EQ(umf_result, UMF_RESULT_SUCCESS);
+    ASSERT_NE(provider, nullptr);
+
+    const char *message;
+    int32_t error = 0;
+    umfMemoryProviderGetLastNativeError(provider, &message, &error);
+    ASSERT_EQ(error, 0);
```
Comment (igchor) [inline test/providers/provider_level_zero.cpp:178]: I don't think you should test for that. The return value of umfMemoryProviderGetLastNativeError is unspecified if the previous call was successful.

File: test/providers/provider_level_zero.cpp [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+
+        // Create context
+        ze_context_desc_t ctxtDesc = {ZE_STRUCTURE_TYPE_CONTEXT_DESC, nullptr,
+                                      0};
+        ze_result = zeContextCreate(hDriver, &ctxtDesc, &hContext);
+        ASSERT_EQ(ze_result, ZE_RESULT_SUCCESS);
+        ASSERT_NE(hContext, nullptr);
+    }
+
+    void TearDown() override { test::TearDown(); }
```
Comment (igchor) [inline test/providers/provider_level_zero.cpp:76]: shouldn't you destroy the context, etc. here?

File: README.md [hunk 12]
```diff
@@ -31,6 +31,10 @@ For development and contributions:
 For building tests, multithreaded benchmarks and Disjoint Pool:
 - C++ compiler with C++17 support
 
+For Level Zero memory provider and its tests:
+- Level Zero headers and libraries
```
[the remark this answers is not in the document]
Comment (bratpiorka) [reply README.md:35]: I've added this requirement one line down - if you have a better idea how to write it, let me know :)

File: src/libumf.map (hunk 2 above)
Comment (bratpiorka) [reply src/libumf.map]: added

File: src/CMakeLists.txt (hunk 4 above)
Comment (bratpiorka) [reply src/CMakeLists.txt]: Yes, but at the moment we don't have the internal infrastructure to test it

File: test/providers/provider_level_zero.cpp (hunk 11 above)
Comment (bratpiorka) [reply test/providers/provider_level_zero.cpp]: thanks, added

File: test/providers/provider_level_zero.cpp (hunk 10 above)
Comment (bratpiorka) [reply test/providers/provider_level_zero.cpp]: removed

File: test/CMakeLists.txt (hunk 9 above)
Comment (bratpiorka) [reply test/CMakeLists.txt]: At the moment we don't have the internal infrastructure to test it. I will try to include Windows ASAP in separate PRs

File: src/provider/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,29 @@
+# Copyright (C) 2024 Intel Corporation
+# Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# libumf_provider_level_zero
+if(UMF_BUILD_LIBUMF_PROVIDER_LEVEL_ZERO)
```
[the remark this answers is not in the document]
Comment (bratpiorka) [reply src/provider/CMakeLists.txt:6]: Yes, but the Level Zero memory provider is a separate library, while the OS memory provider is not (there is only an option to enable/disable it by adding its sources to libumf). If we decide to make it a separate, static library (linked by default), it should be done in a separate PR.

**Review by igchor — APPROVED** [review]
I'd still like to discuss the linking (whether and how we use dlopen) on our sync but the rest looks good.

**Comment by bratpiorka** [comment]
@igchor @vinser52 @pbalcer - please review. I removed dlopen() from the Level Zero provider at all. The test I added to the CI loads libze_loader dynamically using dlopen(RTLD_GLOBAL) and the Level Zero provider reuses these symbols without any problems.

File: src/libumf.map [hunk 14]
```diff
@@ -35,6 +35,7 @@ UMF_1.0 {
         umfMemoryProviderCreateFromMemspace;
         umfMemspaceCreateFromNumaArray;
         umfMemspaceDestroy;
+        umfLevelZeroMemoryProviderOps;
```
[the remark this answers is not in the document]
Comment (bratpiorka) [reply src/libumf.map]: done

File: .github/workflows/gpu.yml [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+          -DUMF_BUILD_LIBUMF_PROVIDER_LEVEL_ZERO=ON
+          -DUMF_BUILD_LIBUMF_POOL_DISJOINT=ON
+          -DUMF_ENABLE_POOL_TRACKING=ON
+
+      - name: Build UMF
+        run: cmake --build ${{github.workspace}}/build -j $(nproc)
+
+      - name: Run tests
+        working-directory: ${{github.workspace}}/build
+        run: |
```
[the remark this answers is not in the document]
Comment (bratpiorka) [reply .github/workflows/gpu.yml]: done

File: include/umf/providers/provider_level_zero.h [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+*/
+
+#ifndef UMF_PROVIDER_LEVEL_ZERO_H
+#define UMF_PROVIDER_LEVEL_ZERO_H
+
+#include <level_zero/ze_api.h>
```
Comment (igchor) [inline include/umf/providers/provider_level_zero.h:11]: I don't think we can include this header here. Our customers might not have the headers installed, if they use dlopen, right? @vinser52?

**Review by igchor — CHANGES_REQUESTED** [review]

**Comment by igchor** [comment]
> @igchor @vinser52 @pbalcer - please review. I removed dlopen() from the Level Zero provider at all. The test I added to the CI loads libze_loader dynamically using dlopen(RTLD_GLOBAL) and the Level Zero provider reuses these symbols without any problems.

Can we also have a similar test for when the application links with Level Zero normally (not using dlopen)? I.e. just have a second test that just assign all those functions pointers in libze_ops to &zeInit, &zeDriverGet, etc?

File: include/umf/providers/provider_level_zero.h (hunk 16 above)
Comment (bratpiorka) [reply include/umf/providers/provider_level_zero.h:11]: but how they could pass ze_context_handle_t or ze_device_handle_t to the level zero provider params?

**Review by lukaszstolarczuk — COMMENTED** [review]
👍

**Comment by bratpiorka** [comment]
> Can we also have a similar test for when the application links with Level Zero normally (not using dlopen)? I.e. just have a second test that just assign all those functions pointers in libze_ops to &zeInit, &zeDriverGet, etc?

@igchor done

File: include/umf/providers/provider_level_zero.h (hunk 16 above)
Comment (igchor) [reply include/umf/providers/provider_level_zero.h:11]: They might have a copy of the headers somewhere, but not necessarily in <level_zero/ze_api.h> I think.

Comment (bratpiorka) [reply include/umf/providers/provider_level_zero.h:11]: or I can check the #define and include the file only when the user hadn't done it yet + print the messafe

Comment (igchor) [reply include/umf/providers/provider_level_zero.h:11]: We can also just accept 'void*' for each param - I don't think there is a good solution for that if we want to use dlopen.

Comment (bratpiorka) [reply include/umf/providers/provider_level_zero.h]: Ok I removed this header from the interface

**Review by igchor — APPROVED** [review]

File: CMakeLists.txt [hunk 17]
```diff
@@ -8,6 +8,7 @@ project(unified-memory-framework VERSION 0.1.0 LANGUAGES C)
 # Build Options
 option(UMF_BUILD_SHARED_LIBRARY "Build UMF as shared library" OFF)
 option(UMF_BUILD_OS_MEMORY_PROVIDER "Build OS memory provider" ON)
+option(UMF_BUILD_LIBUMF_PROVIDER_LEVEL_ZERO "Build Level Zero memory provider" OFF)
```
Comment (vinser52) [inline CMakeLists.txt:11]: if L0 provider implementation has no build-time dependencies why not ON it by default?

File: src/provider/provider_level_zero.c [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+    ze_provider->memory_type = ze_params->memory_type;
+    ze_provider->host_mem_alloc_desc = (ze_host_mem_alloc_desc_t){
+        ZE_STRUCTURE_TYPE_HOST_MEM_ALLOC_DESC,
+        NULL, // pointer to an extension-specific structure
+        ze_params->level_zero_host_mem_alloc_flags};
+    ze_provider->device_mem_alloc_desc = (ze_device_mem_alloc_desc_t){
+        ZE_STRUCTURE_TYPE_DEVICE_MEM_ALLOC_DESC,
+        NULL, // pointer to an extension-specific structure
+        ze_params->level_zero_device_mem_alloc_flags,
+        ze_params->level_zero_device_local_mem_ordinal};
```
Comment (vinser52) [inline src/provider/provider_level_zero.c:93]: I do not understand where it is used and why it is needed?

File: src/provider/provider_level_zero.c [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+static enum umf_result_t ze_memory_provider_alloc(void *provider, size_t size,
+                                                  size_t alignment,
+                                                  void **resultPtr) {
+    assert(provider);
+    assert(resultPtr);
+
+    ze_memory_provider_t *ze_provider = (ze_memory_provider_t *)provider;
+    ze_result_t ze_result = g_ze_ops.zeMemAllocDevice(
```
Comment (vinser52) [inline src/provider/provider_level_zero.c:114]: Based on `ze_provider->memory_type` we should use `zeMemAllocDevice`, `zeMemAllocHost` or `zeMemAllocShared`.

File: src/provider/provider_level_zero.c [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+static enum umf_result_t
+ze_memory_provider_get_min_page_size(void *provider, void *ptr,
+                                     size_t *pageSize) {
+    (void)provider;
+    (void)ptr;
+
+    // TODO
+    *pageSize = 1024;
```
Comment (vinser52) [inline src/provider/provider_level_zero.c:156]: as I know it is 64K

File: src/provider/provider_level_zero.c [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+typedef struct ze_memory_provider_t {
+    ze_context_handle_t context;
+    ze_device_handle_t device;
+    ze_memory_type_t memory_type;
+    ze_host_mem_alloc_desc_t host_mem_alloc_desc;
+    ze_device_mem_alloc_desc_t device_mem_alloc_desc;
+} ze_memory_provider_t;
+
+typedef struct ze_ops_t {
+    ze_result_t (*zeMemAllocDevice)(ze_context_handle_t,
```
[the remark this answers is not in the document]
Comment (bratpiorka) [reply src/provider/provider_level_zero.c:35]: I planned to add shared and host in the next PR which should contain all modes required by UR

File: CMakeLists.txt (hunk 17 above)
Comment (bratpiorka) [reply CMakeLists.txt:11]: no, ze_api.h is still used in src/provider/provider_level_zero.c

File: README.md [hunk 22]
```diff
@@ -123,6 +128,18 @@ A memory provider that provides memory from an operating system.
...
+
+A memory provider that provides memory from a graphics device.
+
+##### Requirements
+
+1) System with Level Zero compatible GPU
+2) Linux OS
+3) The `UMF_BUILD_LIBUMF_PROVIDER_LEVEL_ZERO` option turned `ON`
+4) Required packages:
+   - TODO
```
[the remark this answers is not in the document]
Comment (bratpiorka) [reply README.md:141]: is this a package? I rather thought of libze1 but I'm not sure

File: src/provider/CMakeLists.txt (hunk 13 above)
[the remark this answers is not in the document]
Comment (bratpiorka) [reply src/provider/CMakeLists.txt:6]: Right now I use ze_api.h header but do you suggest removing it completely? I'm not sure if I should re-define Level Zero structures inside UMF

File: src/provider/provider_level_zero.c (hunk 18 above)
Comment (bratpiorka) [reply src/provider/provider_level_zero.c:93]: you have to pass ze_device_mem_alloc_desc_t struct as an argument to zeMemAllocDevice

File: README.md (hunk 22 above)
[the remark this answers is not in the document]
Comment (vinser52) [reply README.md:141]: Ok, I think I got the point. To build L0 provider we need `level-zero-dev` (on Ubuntu, maybe on RedHat it named differently) package. But for distribution, we do not need such dependency, right?

File: CMakeLists.txt (hunk 17 above)
Comment (vinser52) [reply CMakeLists.txt:11]: Ok, there is a build-time dependency. Another question regarding distribution, since L0 provider implementation has no runtime dependency to L0 driver we can make L0 provider implementation to be part of core UMF library.

File: src/provider/CMakeLists.txt (hunk 13 above)
[the remark this answers is not in the document]
Comment (vinser52) [reply src/provider/CMakeLists.txt:6]: but how it relates to the fact that L0 provider is a separate static library?

File: src/provider/provider_level_zero.c (hunk 18 above)
Comment (vinser52) [reply src/provider/provider_level_zero.c:93]: right, but we can create it on stack inside `ze_memory_provider_alloc` function. Furthermore, `device_mem_alloc_desc` is needed only when `ze_provider->memory_type` equals to the device memory - it is another argument not to make it part of the `ze_params` object.

File: CMakeLists.txt (hunk 17 above)
Comment (bratpiorka) [reply CMakeLists.txt:11]: I will make this option enabled by default (ON) + check for build deps in cmake + message to user

File: README.md [hunk 23]
```diff
@@ -123,6 +128,18 @@ A memory provider that provides memory from an operating system.
 4) Required packages for tests:
    - libnuma-dev
 
+#### Level Zero memory provider (Linux-only yet)
+
+A memory provider that provides memory from a graphics device.
```
[the remark this answers is not in the document]
Comment (bratpiorka) [reply README.md]: done

File: README.md (hunk 22 above)
[the remark this answers is not in the document]
Comment (bratpiorka) [reply README.md]: done

File: src/provider/provider_level_zero.c (hunk 21 above)
[the remark this answers is not in the document]
Comment (bratpiorka) [reply src/provider/provider_level_zero.c:36]: added

**Comment by bratpiorka** [comment]
@vinser52 I have resolved the discussions started by you, as we agreed

File: src/provider/provider_level_zero.c (hunk 19 above)
Comment (bratpiorka) [reply src/provider/provider_level_zero.c]: done

File: .github/workflows/pr_push.yml [hunk 24]
```diff
@@ -113,3 +115,6 @@ jobs:
   ProxyLib:
     needs: [Build]
     uses: ./.github/workflows/proxy_lib.yml
+  GPU:
+    needs: [Build]
+    uses: ./.github/workflows/gpu.yml
```
Comment (ldorau) [inline .github/workflows/pr_push.yml:120]: We may add badges for new workflows later ...

**Review by ldorau — CHANGES_REQUESTED** [review]
It does not build on WSL Ubuntu:
```
/home/ldorau/work/unified-memory-framework/src/provider/provider_level_zero.c: In function ‘init_ze_global_state’:
/home/ldorau/work/unified-memory-framework/src/provider/provider_level_zero.c:48:48: error: ‘RTLD_DEFAULT’ undeclared (first use in this function)
   48 |     *(void **)&g_ze_ops.zeMemAllocHost = dlsym(RTLD_DEFAULT, "zeMemAllocHost");
      |                                                ^~~~~~~~~~~~
/home/ldorau/work/unified-memory-framework/src/provider/provider_level_zero.c:48:48: note: each undeclared identifier is reported only once for each function it appears in
make[2]: *** [src/CMakeFiles/umf.dir/build.make:328: src/CMakeFiles/umf.dir/provider/provider_level_zero.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1153: src/CMakeFiles/umf.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

File: .github/workflows/pr_push.yml (hunk 24 above)
Comment (bratpiorka) [reply .github/workflows/pr_push.yml:120]: added

**Comment by bratpiorka** [comment]
@ldorau please check now. BTW we definitely should add WSL to our CI

**Review by ldorau — APPROVED** [review]

**Review by vinser52 — APPROVED** [review]
